### PR TITLE
Initial commit for speculative execution in MR mode

### PR DESCRIPTION
**Pull request by ydai1124**
This is the first PR of https://docs.google.com/document/d/1Vf5A7pHbFKwr7-Bbw2X3zBBVZSvT7HXEPE6bsi9qGco/
Main changes:
1. It has the new `GobblinMultiTaskAttempt`, which does not contain any real execution but provides the layout.
Gobblin `Task` has new `commit` method, which is empty in this PR.
1. It has the new interface `SpeculativeAttemptAwareConstruct`. Some most commonly used gobblin writers, and writer wrappers, have implemented this interface.

@chavdar @ibuenros Can you review?

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java [hunk 1]
```diff
@@ -162,16 +172,20 @@ public long bytesWritten() throws IOException {
...
-  private DataWriter<D> createPartitionWriter(GenericRecord partition) throws IOException {
+  private DataWriter<D> createPartitionWriter(GenericRecord partition)
+      throws IOException {
     if (!this.builder.isPresent()) {
       throw new IOException("Writer builder not found. This is an error in the code.");
     }
-    return this.builder.get().forPartition(partition).withWriterId(this.baseWriterId + "_" + this.writerIdSuffix++)
+    DataWriter dataWriter =  this.builder.get().forPartition(partition).withWriterId(this.baseWriterId + "_" + this.writerIdSuffix++)
         .build();
+    this.isSpeculativeAttemptAware = this.isDataWriterForPartitionSafe(dataWriter);
```
Comment (chavdar) [inline gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java]: You should "and" the safeness as there might be multiple writers instantiated for different partitions.

File: gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java [hunk 2]
```diff
@@ -58,49 +59,55 @@
   private final Optional<PartitionAwareDataWriterBuilder> builder;
   private final boolean shouldPartition;
   private final Closer closer;
+  private boolean isSpeculativeAttemptAware;
```
Comment (chavdar) [inline gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java]: nit: i think the correct name for the variable is "isSpeculativeAttemptSafe"

File: gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    try {
+      List<Future<Void>> futures = new IteratorExecutor<>(callableIterator, this.getTaskCommitThreadPoolSize(),
+          ExecutorsUtils.newDaemonThreadFactory(Optional.of(log), Optional.of("Task-committing-pool-%d"))).execute();
+
+      for (Future<Void> future : futures) {
+        try {
+          future.get();
+        } catch (ExecutionException exc) {
+          log.error("Failed to commit a task.", exc.getCause());
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java]: Shouldn't we fail the whole attempt if one of the tasks fails?

File: gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        }
+      }
+    } catch (InterruptedException ie) {
+      log.error("Committing of tasks interrupted. Aborting.");
+      throw new RuntimeException(ie);
+    }
+
+    for (CommitStep additionalCommitStep: this.additionalCommitSteps) {
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java]: What is the semantics for "additionalCommitSteps" with respect to partial attempt failures?

File: gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java [hunk 5]
```diff
@@ -46,7 +47,7 @@
  * @param <D> record type.
  */
 @Slf4j
-public class PartitionedDataWriter<S, D> implements DataWriter<D>, FinalState {
+public class PartitionedDataWriter<S, D> implements DataWriter<D>, FinalState, SpeculativeAttemptAwareConstruct {
```
Comment (chavdar) [inline gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java]: Please add unit tests for speculative execution awareness of this class

File: gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java (hunk 4 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java]: It is actually the "clean up" steps, which should be always executed. I will put it in a finally cause.

File: gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java (hunk 5 above)
Comment (ydai1124) [inline gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java]: Added.

File: gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java (hunk 1 above)
Comment (ydai1124) [inline gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java]: Fixed.

File: gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java (hunk 3 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java]: Fixed.

**Comment by ydai1124** [comment]
@chavdar Comments addressed.

File: gobblin-api/src/main/java/gobblin/writer/DataWriterBuilder.java [hunk 6]
```diff
@@ -34,6 +34,7 @@
   protected S schema;
   protected int branches;
   protected int branch;
+  protected String writerAttemptId;
```
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/writer/DataWriterBuilder.java]: Do we need a default for this?

File: gobblin-core/src/main/java/gobblin/commit/SpeculativeAttemptAwareConstruct.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * A declaration by any Gobblin construct to claim whether it is safe to have multiple speculative attempts.
+ * For example, if any {@link gobblin.writer.DataWriter} implements {@link SpeculativeAttemptAwareConstruct}
+ * and returns true in {@link #isSpeculativeAttemptSafe()}, then multiple attempts of one {@link gobblin.writer.DataWriter}
+ * should not cause conflict among them.
+ */
+@Alpha
+public interface SpeculativeAttemptAwareConstruct {
+
+  /**
+   * @return true if it is safe to have multiple speculative attempts; false, otherwise.
```
Comment (ibuenros) [inline gobblin-core/src/main/java/gobblin/commit/SpeculativeAttemptAwareConstruct.java]: Can you add to the javadoc the suggested pattern of this? e.g. `return this.class == MyClass.class`

File: gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                  throws Exception {
+                task.commit();
+                return null;
+              }
+            };
+          }
+        });
+
+    try {
+      List<Future<Void>> futures = new IteratorExecutor<>(callableIterator, this.getTaskCommitThreadPoolSize(),
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java]: You can use `executeAndGetResults` instead, then run `IteratorExecutor.verifyAllSuccessful`. If you want to log the failures, you can use `IteratorExecutor.logFailures`.

File: gobblin-api/src/main/java/gobblin/writer/DataWriterBuilder.java (hunk 6 above)
Comment (ydai1124) [inline gobblin-api/src/main/java/gobblin/writer/DataWriterBuilder.java]: It can be null if not specified when building a writer.

**Comment by ydai1124** [comment]
@ibuenros comments addressed.

**Pull request merged**